posts/inside-rust/2023-11-10-parallel-rustc.md

```diff
@@ -0,0 +1,239 @@
+  that, the staircase effect returns because only one "rustc" thread does LLVM
+  IR generation while seven or more other LLVM threads are active. If the
+  number of threads used by the front-end was changed to 16 the staircase shape
+  would disappear entirely, though in this case the final execution time would
+  barely change.
+
+### Putting it all together
+
+Rust compilation has long benefited from intraprocess parallelism, via Cargo,
+and from interprocess parallelism in the back-end. It can now also benefit from
+interprocess parallelism in the front-end.
+
+Attentive readers might be wondering how intraprocess parallelism and
+interprocess parallelism interact. If we have 20 parallel rustc invocations and
+each one can have up to 16 threads running, might we end up 100s of threads on
+a machine with only 10s of cores, resulting in inefficient execution as the OS
+tries its best to schedule them?
+
+Fortunately no. The compiler uses the [jobserver
+protocol](https://www.gnu.org/software/make/manual/html_node/POSIX-Jobserver.html)
+to limit the number of threads it creates. If a lot of interprocess parallelism
+is occuring, intraprocess parallelism will be limited appropriately, and
+the number of threads will not exceed the number of cores.
+
+## How to use it
+
+As of (XXX: date enabled) the nightly compiler is [shipped with the parallel
+front-end enabled](https://github.com/rust-lang/rust/pull/117435). However,
+**by default it runs in single-threaded mode** and won't reduce compile times.
+Keen users who want to try multi-threaded mode should use the `-Z threads=8`
+option.
+
+This default may be surprising. Why parallelize the front-end and then run it
+in single-threaded mode? The answer is simple: caution. This is a big change!
+The parallel front-end has a lot of new code. Single-threaded mode exercises
+most of the new code, but excludes the possibility of threading bugs such as
+deadlocks that still occasionally affect multi-threaded mode. Parallel
+execution is harder to get right than serial execution, even in Rust. For this
+reason the parallel front-end also won't be shipped in Beta or Stable releases
+for some time.
+
+### Performance effects
+
+When the parallel front-end is run in single-threaded mode, compilation times
+are typically within 2% of the serial front-end, which should be barely
+noticeable.
+
+When the parallel front-end is run in multi-threaded mode with `-Z threads=8`,
+out [measurements on real-world
+code](https://github.com/rust-lang/compiler-team/issues/681) on show that
+compile times can be reduced by up to 50%, though the affects vary widely and
+depend greatly on the characteristics of the code being compiled and the build
+configuration. For example, dev builds are likely to see bigger improvements
+than release builds because release builds usually spend more time doing
+optimizations in the back-end. A small number of cases compile more slowly in
+multi-threaded mode than single-threaded mode.
```

Comment:
I think we can mention that most of the cases that get regressions are those crates take small number of time to compile. For example, `hello world` and codes under `incr-unchanged` scenarios.